### PR TITLE
Feature/lazy load abilities

### PR DESCRIPTION
**Pull request by christianblos**
Currently it is not possible to pass dependencies from the bot to an AbilityExtension. Let's take this simple example:

```
public class MyBot extends AbilityBot {
    private final MyConfig config;

    public MyBot(MyConfig config) {
        super("token", "username");
        this.config = config;
    }

    public AbilityExtension createSomeExtension() {
        return new SomeExtension(config);
    }

    @Override
    public int creatorId() {
        return 0;
    }
}
```

Accessing the config object in createSomeExtension() throws a NullPointerException, because abilities are being loaded in the constructor of the AbilityBot. `this.config = config;` was not yet executed, so this.config is null at this point.

In this pull request the abilities are not loaded in the constructor anymore. They are loaded lazily when asking for the abilities the first time. First, I thought about introducing an `initialize()` method, but this would break backward compatibility. So, I decided to go with the lazy loading approach. What are your thoughts?

### CODE REVIEW

**Comment by addo37** [comment]
Changes look good to me. It might make sense to perhaps rely on a factory to create the bot and initialize afterwards. However, this means we're okay in saying "here's one approach to creating the bot that wouldn't initialize your abilities properly, and here's another that does".

Perhaps, we could @deprecated the constructors and mirror them into factory methods. Thoughts?

**Comment by christianblos** [comment]
Since the bot is always a custom class implemented by the user of this library, I think a factory is not really possible 🤔 
The params passed to the constructor are still mandatory for the AbilityBot to work, so I personally wouldn't deprecate it. Just registering the abilities must happen at a later point.

If you're fine with breaking backward compatibility, I could also introduce an `onRegister()` method which is called when the bot gets registered (in `TelegramBotsApi.registerBot()`).

```
public void onRegister() {
    registerAbilities();
    initStats();
}
```

**Comment by addo37** [comment]
Thinking back, that was quite the terrible suggestion. Thanks for pointing out the caveat.

The onRegister method sounds great. I think, it could be made so that WebhookBot and LongPollingBot interface both inherit from another parent interface (Bot/TelegramBot) with the shared methods declared. The parent interface could have getBotUsername, getBotToken, and onRegister.

This change would be transparent to the user tho. People who'd be interested in doing more stuff on registration would simply override, call super method then carry on.

**Comment by christianblos** [comment]
Alright, I reverted my old changes and implemented the onRegister method and the main TelegramBot interface.

**Review by addo37 — CHANGES_REQUESTED** [review]
Please modify the wiki (Bot Testing) to reflect these changes.

File: telegrambots-abilities/src/main/java/org/telegram/abilitybots/api/bot/AbilityBot.java [hunk 1]
```diff
@@ -18,45 +18,50 @@
...
+    }
+
+    protected AbilityBot(String botToken, String botUsername) {
+        this(botToken, botUsername, onlineInstance(botUsername));
+    }
+
+    @Override
+    public void onRegister() {
+        super.onRegister();
+    }
```
[the remark this answers is not in the document]
Comment (christianblos) [reply telegrambots-abilities/src/main/java/org/telegram/abilitybots/api/bot/AbilityBot.java:56]: 🤔 I don't know why I added this. I will remove it again

**Comment by christianblos** [comment]
I also added a section for overriding the `onRegister()` method to the wiki

**Comment by rubenlagus** [comment]
Please apply it on dev branch

**Comment by christianblos** [comment]
Target is now the dev branch

**Pull request merged**